Pico rand

andygpz11 opened the pull request:
Add improved random number generator for RP2040

peterharperuk commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,283 @@
+    // If the RAM has not been hashed or a (re)hash is required, (re)hash the RAM
+    if (!ram_is_hashed || force_ram_rehash) {
+
+        // Note: By design, do *not* initialise the static 'ram_hash' variable
```
won't ram_hash be initialised to zero?

andygpz11 replied:
Ah, it probably will come to think of it :)  So the comment is wrong, I will fix it.

peterharperuk reviewed (commented):
Looks good to me

andygpz11 commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,283 @@
+}
+
+
+// Note: This takes approximately 80 milliseconds to execute!
```
I'm happy to take it out although this code will need "adjusting" or even removing for any future devices _either_ with a (T)RNG _and/or_ a different amount of RAM.  So this is pretty much RP2040 only anyway.
What do you think?

andygpz11 commented:
Comments fixed and pushed (I thought).

kilograham commented:
also please fix compiler warnings/errors failing the build...

just rebuild everything from the pico-sdk SDK repo.

andygpz11 commented:
Yes, of course.  Sorry -  I was _not_ expecting build failures based on everything building here...
I will go round the loop again tomorrow - I have a number of different branches of pico-examples and pico-sdk I need to rationalise (one reason that I want this merged, once - it is correct of course).
Should I be using:
https://github.com/raspberrypi/pico-examples/develop
and
https://github.com/raspberrypi/pico-sdk/develop
Thanks

kilograham commented:
yes develop against the `develop` branches! I rebased your PR onto develop here already

alastairpatrick commented:
I think the intent of not explicitly initializing ram_hash is to get 64 new random bits after each power cycle? In which case, could do:
```static uint64_t ram_hash __attribute__((section(".uninitialized_data")));```
As written, it will go in the BSS section by default and be initialized to zero.

Although the system clock will commonly be derived from the crystal oscillator when measure_rosc_freq() and update_rosc_samples() are called, the system clock could potentially be derived from the relaxation oscillator, in which case the results will not be random. It might be worth panicking or asserting if the system clock is not suitable, unless the other sources of randomness are adequate on their own.

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,88 @@
+ * RAM contents again, not doing so reduces the re-seeding time to ~1ms.
+ *
+ * NOTE:  Code within the rand module will automatically call this function
+ * the first time a RN is requested *if* it has not previously been called.
```
Presumably 'RN' here stands for Random Number? Might be better to avoid the abbreviation, for the sake of clarity.

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,88 @@
+ * be unexpected at that point in program execution.
+ *
+ * \param user_seed  NULL _OR_ pointer to user seed of length PICO_RAND_SEED_SIZE_BYTES
+ * \param force_ram_rehash  If true, *re*hash RAM contents and incorporate in new reed
```
reed -> seed ? :wink:

andygpz11 commented:
Thanks for all the feedback, very useful and now, all I _think_, fixed*
@alastairpatrick I've moved both ram_hash and also the actual 128-bit state used by the XORshift algo uninitialised data, that will offer a useful improvement on reboots - thanks.
*I get your comment about clock source; I'm not sure of the best answer.  The unpredictability of when this gets called, even with the same clock, should still provide a chance of randomness from that source and there are still the temporal aspects provided by system time and hopefully the fact we don't know where the PRNG machine started in the first place.
My feeling is not to panic/assert since that doesn't help the thing work.  
Arguably we need a large disclaimer somewhere say the entropy of the system _might_ be reduced if we aren't using a separate clock.  We offer no guarantee of anything, just scrape round for everything that _might_ add to it.
The one thing in our favour is that entropy is cumulative.  If you incorporate* sources in the correct way you may or may not add entropy but you won't reduce it.  cf: Almost everything else I've seen in engineering degrades with more handling (precision, purity, spectral growth in RF systems) and you can never get it back.
*Obviously collating poor entropy sources doesn't a good RNG make but the maths here is designed to make it harder to work that out remotely. 
Summary: I'm open to suggestions about the best fix.
@kilograham Sorry - I had not spotted the fact that 'kitchensink' ran tests to a higher stander than just forcing the SDK to get remade in support of example code etc - the things that reported were all useful as well, now fixed I hope.

alastairpatrick commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,293 @@
+    rosc_freq = frequency_count_rosc();
+
+    // restore rosc (reduces power)
+    rosc_hw->ctrl = snapshot;
```
This can could cause to clock glitch, which could be an issue if ROSC is the source for the system clock.

alastairpatrick commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,293 @@
+    // else - deliberately leave what is already in 'rand_storage'
+
+    // Note! The safety of the length assumption here is protected by a 'static_assert' above
+    pico_get_unique_board_id(&unique_id.board_id_native);
```
This is less code (the unique_id_u_t union declaration isn't needed) and doesn't rely on a GCC's type punning extension for unions.
```pico_unique_board_id_t unique_id;
pico_get_unique_board_id(&unique_id);
uint64_t unique_id_int;
memcpy(&unique_id_int, &unique_id, sizeof(unique_id_int));
rand_storage.s[0] ^= splitmix64(unique_id_int);```

alastairpatrick commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,293 @@
+}
+
+
+void pico_initialise_rand(void *user_seed, bool force_ram_rehash) {
```
I think this should be:
```void pico_initialise_rand(const struct prng_128_t* user_seed, bool force_ram_rehash)```

`const` because `*user_seed` is not modified. `struct prng_128_t*` rather than `void*` to prevent accidental type punning.

andygpz11 commented:
I have just pushed updated code the this branch, I _think_ this should address the outstanding review comments

peterharperuk commented:
That "Merge branch" commit is a bit odd?

andygpz11 commented:
@peterharperuk Yes, probably.  I took it as "I've updated my branch from develop", which is what I intended.

lurch commented:
I think the way people normally tend to do things is to rebase their feature-branch onto `develop` and then force-push, rather than merging the latest version of `develop` into their feature-branch? :shrug:

JamesH65 commented:
Rather than merge dev back to your branch, I do a `get pull --rebase origin develop` on the branch then force push. I think that might be what @lurch is saying?

lurch commented:
@JamesH65 Yes. Except my muscle-memory always does:
```
git fetch origin
git rebase origin/develop
```

Always more than one way to skin the (octo)cat! :octocat:

peterharperuk commented:
I tested this with mbedtls and it seems to be working.

lurch commented on src/rp2_common/hardware_sync/include/hardware/sync.h:
```diff
@@ -70,6 +70,11 @@ typedef volatile uint32_t spin_lock_t;
 #define PICO_SPINLOCK_ID_HARDWARE_CLAIM 11
 #endif
 
+// PICO_CONFIG: PICO_SPINLOCK_ID_RAND, Spinlock ID for Random Number Generator, min=0, max=31, default=11, group=hardware_sync
```
Huh? Says "default=11" here, but then "PICO_SPINLOCK_ID_RAND 12" a couple of lines further down?

andygpz11 replied:
Thanks for the spot.  Now fixed.

andygpz11 commented on src/rp2_common/pico_rand/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+target_include_directories(pico_rand INTERFACE ${CMAKE_CURRENT_LIST_DIR}/include)
+
+target_link_libraries(pico_rand INTERFACE pico_time pico_unique_id)
```
Agreed about time but I do think I need:
target_link_libraries(pico_rand INTERFACE **pico_unique_id**)
As, during testing with a "real" example, the build fails for the lack of it.

kilograham replied:
what i meant was, use the libraries you DO require (which is pretty clear from what you `#include`)

andygpz11 commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,302 @@
+}
+
+
+static uint32_t frequency_count_rosc(void) {
```
Most of the module ends up being called from _get_new_rand_64()_ which applies the lock.
I've tagged any function that are not deterministic in their execution times.

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,74 @@
+ * subsequent random numbers taking between 10 and 20 microseconds to generate.
+ */
+
+// PICO_CONFIG: PICO_RAND_MIN_ROSC_SAMPLE_TIME_US, Define a default minimum time between sampling the ROSC random bit, min=5, max=20, group=pico_rand
```
Just out of curiosity, is there a reason for these particular min/max values?
Probably makes sense to also include the `default=10` that you use in the `.c` file?

andygpz11 replied:
I did a lot of testing using https://github.com/Fourmilab/ent_random_sequence_tester on results collected over a serial port (over many hours : )

And, for whatever reason, observed better results with samples spaced 10us apart of more, than faster samples.
The limits I have specified are something of an estimation, based on the fact that 10us was a decent data point.

Do I have the option not to specify min and max I wonder?

lurch replied:
Yeah, I think the min and max are only so that https://github.com/raspberrypi/pico-project-generator can display suitable options. Should be fine to leave them off if you want.

andygpz11 replied:
Thanks.  I might stiffen the text a tiny bit to indicate 10us was as a result of testing.

andygpz11 commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,303 @@
+
+    // Generate a 64-bit RN from the modified PRNG state.
+    // Note: This also "churns" the 128-bit state for next time.
+    rand64 = xoroshiro128ss();
+
+    spin_unlock(lock, save);
```
It does need to be under lock.  Did you see my comment above?

_Most of the module ends up being called from get_new_rand_64() which applies the lock.
I've tagged any function that are not deterministic in their execution times._

I have no issue with tagging everything under lock which would also include the splitmix64 calls as well of course.

lurch replied:
> I have no issue with tagging everything under lock

That sounds like a Graham-decision to me...

kilograham replied:
yeah i mean for consistency, then yes.

If every static function requires lock, then possibly just documenting the entire group is fine, but I have no problem with being explicit in all function names

andygpz11 replied:
Sorry, how do you mean, "documenting the entire group"?
Just with a comment or is there a "more substantial" mechanism I've missed please?
FYI:  _All_ the functions from the start of the module until  get_new_rand_64()  (where the lock is applied) are locked.
Perhaps that helps explain this.

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,74 @@
+ *
+ * Note: The *first* time a random number is requested, the 128-bit PRNG state
+ * must be seeded.  The seed generation takes approximately 1 millisecond with
+ * subsequent random numbers taking between 10 and 20 microseconds to generate.
```
Presumably this time is also dependent upon what `PICO_RAND_MIN_ROSC_SAMPLE_TIME_US` has been set to?

andygpz11 replied:
Absolutely

bruelltuete commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,303 @@
+
+    // speed up rosc, wait until stable
+    rosc_hw->ctrl = (ROSC_CTRL_ENABLE_VALUE_ENABLE   << ROSC_CTRL_ENABLE_LSB) |
+                    (ROSC_CTRL_FREQ_RANGE_VALUE_HIGH << ROSC_CTRL_FREQ_RANGE_LSB);
```
what does speeding up here do for the random number generator?
does the sped-up frequency only go into initialisation as a tiny bit of startup entropy?
could we make this optional? there might be stuff running from the rosc (other than clk_sys).

kilograham replied:
yeah, i removed this

lurch commented on src/rp2_common/pico_platform/include/pico/platform.h:
```diff
@@ -153,14 +153,14 @@ extern "C" {
- *     uint32_t __uninitialized_ram("my_group_name") foo;
+ *     uint32_t __uninitialized_ram("foo");
  *
  * The section attribute is `.uninitialized_ram.<group>`
```
Does this comment need updating to `.uninitialized_data.<group>` to match the change a few lines below, or am I misunderstanding? :shrug:   (I don't know anything about linker stuff)

kilograham replied:
no that comment was correct, the one i changed wasn;t

lurch commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,237 @@
+#ifndef PICO_RAND_RAM_HASH_END
+#define PICO_RAND_RAM_HASH_END     SRAM_END
+#endif
+// PICO_CONFIG: PICO_RAND_RAM_HASH_START, end of address in RAM rand (non-inclusive) to hash during pico_rand initialization, default=SRAM_END, group=pico_rand
```
"end of address in RAM rand" -> "start of address in RAM rand" ? Also, the `default=` seems incorrect?

lurch commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,237 @@
+// requested from interrupt context only a small amount of RAM should be hashed
+// hence the 1k default here which will take ~300us on RP2040 at 125Mhz
+
+// PICO_CONFIG: PICO_RAND_DISABLE_ROSC_CHECKS, Dsiable assertions that the ROSC is running when generating random numbers which will drastically reduce randomness, type=bool, default=0, group=pico_rand
```
"Dsiable" -> "Disable"

lurch commented on src/rp2_common/pico_platform/include/pico/platform.h:
```diff
@@ -153,14 +153,14 @@ extern "C" {
  *              garbage-collected independently
  */
-#define __uninitialized_ram(group) __attribute__((section(".uninitialized_ram." #group))) group
+#define __uninitialized_ram(group) __attribute__((section(".uninitialized_data." #group))) group
```
@kilograham This is the "change a few lines below" that I was referring to in my previous comment, but like I said I don't really understand all this stuff.

Looking at the other defines in this file, is line 177 missing a `.` off the end of the name, i.e. should `".flashdata"` be `".flashdata."` ? :shrug:

lurch commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,241 @@
+#define PICO_RAND_RAM_HASH_START   (PICO_RAND_RAM_HASH_END - 1024u)
+#endif
+
+// PICO_CONFIG: PICO_RAND_MIN_ROSC_SAMPLE_TIME_US, Define a default minimum time between sampling the ROSC random bit, min=5, max=20, group=pico_rand
```
Should probably add `default=10` here

lurch commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,241 @@
+// requested from interrupt context only a small amount of RAM should be hashed
+// hence the 1k default here which will take ~100us on RP2040 at 125Mhz
+
+// PICO_CONFIG: PICO_RAND_DISABLE_ROSC_CHECKS, Disable assertions that the ROSC is running when generating random numbers which will drastically reduce randomness, type=bool, default=0, group=pico_rand
```
This comment _might_ be misread that simply disabling assertions reduces randomness? :wink: 
Could perhaps be something like "Disable assertions that the ROSC is running when generating random numbers. Not having the ROSC running will drastically reduce randomness" ?

lurch replied:
And does this need an explicit
```
#ifndef PICO_RAND_DISABLE_ROSC_CHECKS
#define PICO_RAND_DISABLE_ROSC_CHECKS 0
#endif
```
Or is that superfluous for (boolean) PICO_CONFIG items that default to 0?

kilograham replied:
it is now undocumented, as not strictly needed

lurch commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,241 @@
+    // Ensure that the ROSC random bit is not sampled too quickly,
+    // ROSC may be ticking only a few times a microsecond.
+    // Note: In general (i.e. sporadic) use, very often there will be no delay here.
+    while ((time_us_32() - previous_sample_time_us) < PICO_RAND_MIN_ROSC_SAMPLE_TIME_US) {
```
If the ROSC **isn't** enabled, is there any point `in having this delay? (Or even in calling `update_rosc_samples_under_lock` at all?)  But perhaps tweaking that is an optimisation that's not worth it.

kilograham replied:
Actually we need to fix this; this is sleeping with interrupts disabled

P33M commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -140,100 +116,186 @@ static uint64_t sdbm_hash64_sram(uint64_t hash) {
+    static absolute_time_t next_sample_time;
 
-    check_rosc_asserts();
+    // provide an override if someone really wants it, but disabling ROSC as an entroy source makes more sense
```
Typo: `entropy`

P33M commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -18,23 +18,128 @@ extern "C" {
+ *  - Time (\ref PICO_RAND_SEED_ENTROPY_SRC_TIME == 1): The 64-bit microsecond timer is mixed into the seed.
+ *  - Board Identifier (PICO_RAND_SEED_ENTROPY_SRC_BOARD_ID == 1): The board idea via \ref pico_get_unique_board_id
+ *    is mixed into the seed.
+ *  - RAM has (\ref PICO_RAND_SEED_ENTROPY_SRC_RAM_HASH (\ref PICO_RAND_SEED_ENTROPY_SRC_RAM_HAS): The hashed contents of a
```
Typo: `RAM hash`

lurch replied:
Also `PICO_RAND_SEED_ENTROPY_SRC_RAM_HAS`

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,173 @@
+ *  - Bus Performance Counter (\ref PICO_RAND_ENTROPY_SRC_BUS_PERF_COUNTER == 1): One of the bus fabric's performance
+ *    counters is mixed in each time.
+ *
+ * \note all entropy sources are hashed before application to the PRNG state machine.
```
Minor nitpick, but should probably be a capital 'A' on 'all'.

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,173 @@
+ *  - The Ring Oscillator (ROSC) (\ref PICO_RAND_SEED_ENTROPY_SRC_ROSC == 1):
+ *    64 bits are gathered from the ring oscillator "random bit" and mixed into the seed.
+ *  - Time (\ref PICO_RAND_SEED_ENTROPY_SRC_TIME == 1): The 64-bit microsecond timer is mixed into the seed.
+ *  - Board Identifier (PICO_RAND_SEED_ENTROPY_SRC_BOARD_ID == 1): The board idea via \ref pico_get_unique_board_id
```
"board idea" -> "board id"

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,173 @@
+#define PICO_RAND_SEED_ENTROPY_SRC_TIME PICO_RAND_ENTROPY_SRC_TIME
+#endif
+
+// PICO_CONFIG: PICO_RAND_SEED_ENTROPY_SRC_BOARD_ID, Enable/disable use of as part of the random seed, type=bool, default=1, group=pico_rand
```
Missing word(s) between "of" and "as" ?

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,173 @@
+#define PICO_RAND_ROSC_BIT_SAMPLE_COUNT 1
+#endif
+
+// PICO_CONFIG: PICO_RAND_MIN_ROSC_BIT_SAMPLE_TIME_US, Define a default minimum time between sampling the ROSC random bit, min=5, max=20, group=pico_rand
```
Should add `default=10,` here

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,173 @@
+// PICO_RAND_SEED_ENTROPY_SRC_RAM_HASH CONFIG
+// ------------------------------------------
+
+// PICO_CONFIG: PICO_RAND_RAM_HASH_END, end of address in RAM (non-inclusive) to hash during pico_rand SEEDization, default=SRAM_END, group=pico_rand
```
Fairly sure that "SEEDization" isn't a proper word - I dunno if this was deliberate or is a search'n'replace error?

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,173 @@
+#ifndef PICO_RAND_RAM_HASH_END
+#define PICO_RAND_RAM_HASH_END     SRAM_END
+#endif
+// PICO_CONFIG: PICO_RAND_RAM_HASH_START, start of address in RAM (inclusive) to hash during pico_rand SEEDization, default=PICO_RAND_RAM_HASH_END-1024, group=pico_rand
```
I _guess_ you could add MIN and MAX values for the `PICO_RAND_RAM_HASH_START` and `PICO_RAND_RAM_HASH_END` PICO_CONFIG entries?

kilograham replied:
nah

lurch commented on src/rp2_common/pico_rand/include/pico/rand.h:
```diff
@@ -0,0 +1,173 @@
+// PICO_RAND_ENTROPY_SRC_BUS_PERF_COUNTER CONFIG
+// ---------------------------------------------
+
+// PICO_CONFIG: PICO_RAND_BUS_PERF_COUNTER_INDEX, Bus performance counter index to use for sourcing entropy, group=pico_rand
```
`min=0, max=3` ?

lurch commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,303 @@
+    static volatile uint16_t check_byte;
+    rng_128_t local_rng_state = rng_state;
+    uint8_t local_check_byte = check_byte;
+    // Modify PRNG state with the two run-time entropy sources,
```
"the two run-time entropy sources" -> "the run-time entropy sources"

lurch commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,303 @@
+    }
+    local_rng_state.r[which] &= splitmix64(bus_counter_value);
+    which ^= 1;
+#endif
```
Should there be a compile-time warning/error if PICO_RAND_ENTROPY_SRC_TIME, PICO_RAND_ENTROPY_SRC_ROSC and PICO_RAND_ENTROPY_SRC_BUS_PERF_COUNTER are **all** disbaled? Or is that too much hand-holding?

kilograham replied:
agreed

lurch commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,303 @@
+        initialise_rand();
+    }
+
+    static volatile uint16_t check_byte;
```
`uint8_t`, to match local_check_byte ?

lurch commented on src/rp2_common/pico_rand/rand.c:
```diff
@@ -0,0 +1,303 @@
+*/
+
+#include "pico/rand.h"
+#include "pico/unique_id.h"
```
Is it worth making these includes conditional, based upon what entropy sources have been selected?  e.g. could have
```
#if PICO_RAND_SEED_ENTROPY_SRC_BOARD_ID
#include "pico/unique_id.h"
#endif
```

kilograham replied:
yeah, i thought about it, but is more clutter, and only really helps if we remove the requirement in the CMakeLists.txt dependencies (where we don't have the `#define`
